source/distributed/fully_distributed_tria.cc

```diff
@@ -0,0 +1,525 @@
+             const double                  right,
+             const bool                    colorize);
+} // namespace GridGenerator
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    namespace internal
+    {
+      /**
+       * Check if `parent` is the parent of `child`.
+       */
+      template <int dim>
+      bool
+      is_parent(typename CellId::binary_type parent,
+                typename CellId::binary_type child)
+      {
+        if (child[0] != parent[0])
+          return false; // has same coarse cell
+        if ((child[1] >> 2) != ((parent[1] >> 2) + 1))
+          return false; // level has one difference
+
+
+        const unsigned int n_child_indices = parent[1] >> 2;
+        const unsigned int children_per_value =
+          sizeof(CellId::binary_type::value_type) * 8 / dim;
+        unsigned int child_level  = 0;
+        unsigned int binary_entry = 2;
+
+        // Loop until all child indices have been written
+        while (child_level < n_child_indices)
+          {
+            Assert(binary_entry < child.size(), ExcInternalError());
+
+            for (unsigned int j = 0; j < children_per_value; ++j)
+              {
+                if ((((child[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)) !=
+                    (((parent[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)))
+                  return false;
+
+
+                ++child_level;
+                if (child_level == n_child_indices)
+                  break;
+              }
+
+            ++binary_entry;
+          }
+        return true;
+      }
+    } // namespace internal
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::reinit(
+      const ConstructionData<dim, spacedim> &construction_data)
+    {
+      // clear internal data structures
+      this->coarse_cell_id_to_coarse_cell_index_vector.clear();
+      this->coarse_cell_index_to_coarse_cell_id_vector.clear();
+
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.coarse_cell_vertices.empty() == true)
+        {
+          // 1) create a dummy hypercube
+          create_triangulation_for_internal_usage = true;
+          GridGenerator::hyper_cube(*this, 0, 1, false);
+          create_triangulation_for_internal_usage = false;
+
+          // 2) mark cell as artificial
+          auto cell = this->begin();
+          cell->set_subdomain_id(dealii::numbers::artificial_subdomain_id);
+          cell->set_level_subdomain_id(
+            dealii::numbers::artificial_subdomain_id);
+
+          // 3) setup dummy mapping between locally relevant coarse-grid cells
+          //    and global cells
+          this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(
+            numbers::invalid_coarse_cell_id, 0);
+          this->coarse_cell_index_to_coarse_cell_id_vector.push_back(
```

Comment:
Since you use `emplace_back` above, just use it here as well (even though it doesn't make a difference)